### ISSUE TITLE
[Question] Error messages from vm-import-controller

### ISSUE THREAD

**Issue by mjj29** [harvester/harvester#5525]
I'm trying to test vmware migration into harvester using the vm-import-controller. I've added it and configured the import controller, and now I don't get any error messages from the controller pod, but it still reports as cluster not ready:

    $ kubectl get vmwaresource.migration
    NAME    STATUS
    vcsim   clusterNotReady`

here are the import controller logs, even after restarting the pod:

    $ kubectl logs -n harvester-system  pod/vm-import-controller-harvester-vm-import-controller-986f5ffm2v5
    time="2024-04-09T13:22:10Z" level=info msg="Applying CRD vmwaresources.migration.harvesterhci.io"
    time="2024-04-09T13:22:11Z" level=info msg="Applying CRD openstacksources.migration.harvesterhci.io"
    time="2024-04-09T13:22:11Z" level=info msg="Applying CRD virtualmachineimports.migration.harvesterhci.io"
    time="2024-04-09T13:22:12Z" level=info msg="Starting migration.harvesterhci.io/v1beta1, Kind=VirtualMachineImport controller"
    time="2024-04-09T13:22:12Z" level=info msg="Starting harvesterhci.io/v1beta1, Kind=VirtualMachineImage controller"
    time="2024-04-09T13:22:12Z" level=info msg="Starting migration.harvesterhci.io/v1beta1, Kind=VmwareSource controller"
    time="2024-04-09T13:22:12Z" level=info msg="Starting migration.harvesterhci.io/v1beta1, Kind=OpenstackSource controller"  
    time="2024-04-09T13:22:12Z" level=info msg="reoncilling vmware migration default/vcsim"

When I add a VM to import I get this:

    time="2024-04-09T13:28:13Z" level=error msg="error syncing 'apama/bcv-rhel7-import-test': handler virtualmachine-import-job-change: migration not yet ready. current status is clusterNotReady, requeuing"

Is there any way to get any more information as to why it's not working?

**Comment by ibrokethecloud** [comment]
the source is not ready. Likely because of incorrect datacenter specification in the vmware source.

We can definitely improve the logging in the controller to help identify these issues better.

**Comment by mjj29** [comment]
I agree that the source is probably not ready, but how do I know _why_ it's not ready. The dc listed matches the name I see in vsphere, so it should be correct, but how do I know?

**Comment by mjj29** [comment]
Hey, any other thoughts?

**Comment by mjj29** [comment]
Hey, any other thoughts?

**Comment by ibrokethecloud** [comment]
are you able to update the addon to use the latest image `rancher/harvester-vm-import-controller:main-head` this has additional changes to print the actual error when check fails.

**Comment by mjj29** [comment]
OK, I've done that and I get:
time="2024-05-01T08:43:27Z" level=error msg="failed to verfiy client for vmware migration" apiVersion=migration.harvesterhci.io/v1beta1 err=NoPermission kind=VmwareSource name=vcsim namespace=default

**Comment by mjj29** [comment]
I've reviewed the configuration and the secrets and they appear correct

**Comment by mjj29** [comment]
Hey, any update?

**Comment by ibrokethecloud** [comment]
the error indicates the account used does not have permissions to list the datacenter you have specified in the source.

**Comment by mjj29** [comment]
We've confirmed that my account has Administrator access to the datacenter and child objects

**Comment by ibrokethecloud** [comment]
any chance i could get a support bundle from this cluster

**Comment by mjj29** [comment]
Right, I've gotten further. We now have a source which is ready:

    $ rancher kubectl  get vmwaresource.migration
    NAME    STATUS
    vcsim   clusterReady

This is using this spec:

   endpoint: "https://daevcsa01.eur.ad.sag/sdk"
   dc: ""

(I've tried various values for dc, see below)

Now I try and migrate a VM and I get this:

    handler virtualmachine-import-job-change: error in poweroff call: error finding vm in PowerOffVirtualMachine: vm '/vm/PAM Build/vm-646554' not found, requeuing"

I have tried a lot of things in the DC, and the folder and the VM name to try and get it to work, but everything gives me that error. Let me tell you what I have in my setup in vsphere and in the rest API. Notably, the rest API I use is not /sdk, we have no idea what that is, it's with /rest/vcenter.

In the vsphere UI I see:

![image](https://github.com/harvester/harvester/assets/9658120/d6cafdc1-d520-412c-84d6-5d4127c48823)
![image](https://github.com/harvester/harvester/assets/9658120/9decee8f-d030-4189-8023-aaca5e575eb6)

When I query the rest API I get:

curl -X GET https://daevcsa01.eur.ad.sag/rest/vcenter/vm/

          { 
             "power_state" : "POWERED_ON",                                                               
             "name" : "abv-rhel7",
             "cpu_count" : 4,                                                                            
             "vm" : "vm-646554",  
             "memory_size_MiB" : 8192                                                                 
           },   

This then allows me to get details via: curl -X GET https://daevcsa01.eur.ad.sag/rest/vcenter/vm/vm-646554

I have tried changing dc, folder and virtualMachineName so that the following paths have all been reported as failed:

* /CUK/vm/abv-rhel7
* /CUK/Apama/vm/abv-rhel7
* /CUK/vm/Apama/abv-rhel7
* /CUK/vm/PAM Build/abv-rhel7
* /vm/abv-rhel7
* /vm/Apama/abv-rhel7

And most of the same with abv-rhel7 replaced with vm-646554

What would you expect me to put in DC, folder and virtualMachineName for this to work? What path would you expect to see in that log message? How can I verify in the UI or the REST API which is correct?

Thanks,
Matt

**Comment by mjj29** [comment]
Hey, any thoughts on this?

**Comment by mjj29** [comment]
Hey, any ideas? 
Thanks
Matt

**Comment by mjj29** [comment]
Hey there, I think we're really close, it would just be great to have some thoughts from someone on what this _ought_ to look like from my last message.

Thanks,
Matt

**Comment by mjj29** [comment]
Hey, I don't suppose anyone has had a chance to have any thoughts on this?
Thanks,
Matt

**Comment by kevinmata92** [comment]
Also running to this issue in our environment. Bumping thread to get more info

**Comment by mjj29** [comment]
Hi there, it would be really helpful if anyone has some time to look at the latest info I posted and see if they have any suggestions as to what I'm doing wrong, thanks

**Comment by zanewgray** [comment]
I hate to be the "Me too." poster, but I'm having the same issue.  

I was encouraged enough by https://github.com/harvester/harvester/issues/5143 that I even rebuilt my Harvester cluster on 1.3.1, but the logs on VM import still show:

`handler virtualmachine-import-job-change: error in poweroff call: error finding vm in PowerOffVirtualMachine: vm`

I've tried several possible combinations of VCenter paths, as well as using the `administrator` account for the VMWare cluster.  I've manually powered off the VM in hopes of overcoming the `PowerOffVirtualMachine` error, and renamed the VM due to it originally having uppercase characters (which the migration CRD balks at).  `curl` commands run from the Harvester nodes confirm connectivity between Harvester and VCenter, so this does not appear to be a network path issue.

I'll gladly provide a support bundle if that helps.  Thanks!

**Comment by zanewgray** [comment]
Just a follow up thought: the VM Import enhancement spec appears to allow for a `path` variable in the CRD:

https://github.com/harvester/harvester/blob/master/enhancements/20220726-vm-migration.md

The VM that I am attempting to migrate is within a folder structure:

![image](https://github.com/harvester/harvester/assets/16705219/cdf72daa-6b42-4c74-a01e-07505f695646)

However, the final spec for key name related to a `path` variable did not make it into the documented example (ref: https://docs.harvesterhci.io/v1.3/advanced/addons/vmimport)  So it might not be `path`, but some variation (e.g. `virtualMachinePath` or the like), and I am unable to find any other example CRDs that leverage a path. 

In short, what is the proper means of defining a VM `path` variable into the CRD?

**Comment by zanewgray** [comment]
Getting closer!  

Reading the code at https://github.com/harvester/vm-import-controller/blob/main/pkg/apis/migration.harvesterhci.io/v1beta1/virtualmachines.go, I found that the CRD spec to define the path is `folder`.  My CRD now looks like:

```
apiVersion: migration.harvesterhci.io/v1beta1
kind: VirtualMachineImport
metadata:
  name: struxurewaredcexpert
  namespace: default
spec:
  virtualMachineName: "struxurewaredcexpert"
  folder: "/zgray/Lab-Tools"
  networkMapping:
  sourceCluster:
    name: vc-bar
    namespace: default
    kind: VmwareSource
    apiVersion: migration.harvesterhci.io/v1beta1
```

However, looking at the migration pod logs, I'm seeing a character limit that I'm bumping into:

`time="2024-06-24T19:33:21Z" level=error msg="error syncing 'default/struxurewaredcexpert': handler virtualmachine-import-job-change: error creating vmi: VirtualMachineImage.harvesterhci.io \"image-tz2lv\" is invalid: metadata.labels: Invalid value: \"vm-import-struxurewaredcexpert-struxurewaredcexpert-default-disk-0.img\": must be no more than 63 characters, requeuing"`

This appears to be a documented bug (ref: https://github.com/harvester/harvester/issues/5717)

Regardless, this should resolve @mjj29 problem as their path ('folder') + VM name "should" fall below the 63 character issue (fingers crossed...)

**Comment by zanewgray** [comment]
Final update:

I adjusted the metadata -> name value from `struxurewaredcexpert` to `dcexpert`.  This overcame the 63 character limit issue.

Once I re-applied the CRD, Harvester requested that VCenter power down the VM (was in a paused state) and export an OVF template of the VM to Harvester.  Harvester then imported the volume, created the VM and, powered it on.

There is still likely a need to install VirtIO drivers on the VM, but, otherwise, this is a success!

For the Suse folks reading this, my only feedback would be to update the documentation to describe the available key/values that the CRD can leverage.  A simple chart of the values, along with some follow-on examples, would suffice.  Knowing about the existence of a `folder` specification would have saved me some time and frustration.

Realizing that the enhancement spec mentions incorporating this into the UI (maybe add this to the roadmap?), I'll reserve my end-user experience feedback for now.

**Comment by matj-sag** [comment]
Aha, thanks. I did also discover folder, but it wasn't working for me. Turns out I wasn't trying with a leading /. I think that was working, but now I'm getting auth errors, so I need to check that on my side. Examples using folder would be really helpful

**Comment by mjj29** [comment]
Hi all, so I think @zanewgray 's answer fixes the problem I was previously having, but now I'm in the position where I get clusterReady on the cluster, and sourceReady on the VM, but then nothing happens with nothing in the import controller logs:

    NAME   STATUS
    cuk    clusterReady

VM

    NAME             STATUS
    abv-rhel7-test   sourceReady
    
Logs:

    time="2024-07-29T09:15:55Z" level=info msg="Applying CRD vmwaresources.migration.harvesterhci.io"
    time="2024-07-29T09:15:55Z" level=info msg="Applying CRD openstacksources.migration.harvesterhci.io"
    time="2024-07-29T09:15:55Z" level=info msg="Applying CRD virtualmachineimports.migration.harvesterhci.io"
    time="2024-07-29T09:15:56Z" level=info msg="Starting migration.harvesterhci.io/v1beta1, Kind=OpenstackSource controller"
    time="2024-07-29T09:15:56Z" level=info msg="Starting migration.harvesterhci.io/v1beta1, Kind=VirtualMachineImport controller"
    time="2024-07-29T09:15:56Z" level=info msg="Starting harvesterhci.io/v1beta1, Kind=VirtualMachineImage controller"
    time="2024-07-29T09:15:56Z" level=info msg="Starting migration.harvesterhci.io/v1beta1, Kind=VmwareSource controller"
    time="2024-07-29T09:15:56Z" level=info msg="reoncilling vmware migration default/cuk"

What's going on?

**Comment by mjj29** [comment]
OK, follow up - "sourceReady" means "in progress". After some time, it went to virtualMachineCreated and my VM booted (yay!), but the disk is degraded and it dropped into init safe mode with no disks (boo!). Longhorn is reporting it can't schedule the replica, saying there's no disk available, despite having 2TB free above the reservation (single node, 2TB data disk, just deleted all the other volumes):

    time="2024-07-29T09:45:50Z" level=error msg="There's no available disk for replica pvc-0ccd9955-e6ed-4a8d-9fd4-b8c9f0445a5c-r-4942bc66, size 107374182400" func="scheduler.(*ReplicaScheduler).ScheduleReplica" file="replica_scheduler.go:95"
    time="2024-07-29T09:45:50Z" level=warning msg="Failed to schedule replica" func="controller.(*VolumeController).reconcileVolumeCondition" file="volume_controller.go:1694" accessMode=rwx controller=longhorn-volume frontend=blockdev migratable=true node=iotperf09 owner=iotperf09 replica=pvc-0ccd9955-e6ed-4a8d-9fd4-b8c9f0445a5c-r-4942bc66 shareEndpoint= shareState= state=attached volume=pvc-0ccd9955-e6ed-4a8d-9fd4-b8c9f0445a5c

**Comment by mjj29** [comment]
As, so I'm testing this in a single-node cluster and the import creates volumes with replicas=3. My actual deployment will have 5 nodes, so this will be fine. However, even after fixing that I still get the VM dropping into safe mode and it doesn't see any disks. This is a rhel7 VM, so I wonder if there are any driver issues, and how I would fix that. I'm trying again with a rhel8 VM and a win2019 VM to see if they import OK

**Comment by zanewgray** [comment]
One thought: Prior to importing the VM, log into the VM on the source system and install the appropriate Qemu/VirtIO drivers.  This "might" allow it to boot after the import.

I also seem to recall (but can't seem to find it at the moment) either an enhancement or some instructions on how to change the drive type after the import (e.g. VirtIO to IDE or SCSI).  This might help (assuming you can find it) if installing drivers before the import doesn't resolve your issue.

Thanks for sharing your notes!

**Comment by mjj29** [comment]
And with a change to prep the VM to include the virtio drivers in the initramfs before migration, I have a successfully migrated linux VM. I'm testing a windows VM now, but I think it's likely this will be working finally. Thanks for your help all.

I also ran out of space on my import controller drive, but I figured out how to change that as well.

**Comment by zanewgray** [comment]
One other note for anyone who reads this in the future: If your source system is VSphere/VCenter, you may also have to remove the VMWare Tools from the VM prior to a V2V migration (be sure to snapshot/backup the VM prior to doing this).  Use the VCenter console to do this as using RDP/VNC/SSH to the VM itself will likely get cut off during the uninstall.

Background: I ran into an issue on a different V2V operation (between VCenter and HyperV) where having the VMWare Tools installed cause the NIC to not come up or grab a DHCP lease due to the VM trying to use a (no longer in existence) OpenVSwitch component.  I suspect that a V2V between VCenter and Harvester "may" result in a VM with similar issues.  

I did resolve this on an Ubuntu 22.04 migrated VM using the following steps:

1. Add an `/etc/netplan/99_config.yaml` file to the VM.
2. Add these contents to the file:

```
network:
  version: 2
  renderer: networkd
  ethernets:
    eth0:                              <-- you can use "ip a" to determine your adapter name
      dhcp4: true
```

3. Apply the config with `sudo netplan apply`.  The VM grabbed a DHCP lease on the configured VLAN.
4. You "may" have to also perform a `sudo ip link set eth0 up` as well (was down on first boot before the changes)

**Comment by mjj29** [comment]
OK, I've run into a further issue - when I try migrating my windows machine it times out after 2 hours. Then it restarts. Then it timed out again

**Comment by zanewgray** [comment]
@mjj29 silly question: is your VM import controller PVC/drive at least twice the size of the VM that you are trying to migrate?  From the docs:

"During the migration, a large VM's node could run out of space on this mount, resulting in subsequent scheduling failures.

To avoid this, users are advised to enable PVC-backed storage and customize the amount of storage needed. According to the best practice, the PVC size should be twice the size of the largest VM being migrated. This is essential as the PVC is used as scratch space to download the VM, and convert the disks into raw image files."

**Comment by mjj29** [comment]
Yeah, I have 800G on the import controller drive (I ran into disk space issues already and had to resize it), I'm fairly sure it's a timeout

**Comment by mjj29** [comment]
Hmm, have you tried migrating any windows machines? How do you get the virtio drivers installed so that it boots after the migration? I ran the guest agent and driver installation before the migration, but it still can't find them and won't boot

**Comment by zanewgray** [comment]
Unfortunately, my use cases involve Linux distributions.  I have not tried migrating a Windows-based VM.

Referencing https://github.com/harvester/harvester/issues/5143 - I did see that this was noted in the tread:

"For Windows, there is the need to Install the SUSE Virtual Machine Driver pack (virtIO drivers) or utilize SCSI/SATA."

So, maybe editing the VM after import to change the volume bus type might get it to a booting state.

**Comment by mjj29** [comment]
OK, for windows you have to switch to sata for the disk and e1000 for the network. I tried installing the virtio drivers in advance, but no luck

**Comment by mjj29** [comment]
Hmm, Now I'm having issues with rhel7. I have the virtio drivers in the initrd, but it's still not detecting the disk, and it's not working with sata either

**Comment by Ansible-man** [comment]
We are also having issues with the path and the logs showing that the VM cannot be found. In referencing the most recent documentation we found nothing useful.
